password_enhancements.module

```diff
@@ -0,0 +1,484 @@
+        unset($constraints[$constraint_id]);
+      }
+    }
+
+    $checked_constraints += $usable_constraints_by_policy;
+  }
+
+  return $constraint_list;
+}
+
+/**
+ * Ajax callback for updating the constraints on role change.
+ *
+ * @param array $form
+ *   The form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current form state.
+ *
+ * @return array
+ *   Renderable form array for password policy constraints.
+ */
+function password_enhancements_update_constraints_by_roles(array $form, FormStateInterface $form_state): array {
+  return $form['account']['password_policy_constraints'];
+}
+
+/**
+ * Implements hook_entity_base_field_info().
+ */
+function password_enhancements_entity_base_field_info(EntityTypeInterface $entity_type): array {
+  $fields = [];
+
+  if ($entity_type->id() == 'user') {
+    $fields['password_enhancements_password_changed_date'] = BaseFieldDefinition::create('datetime')
+      ->setLabel(t('Last password change'))
+      ->setDescription(t('The last time when the user has changed the password.'))
+      ->setDefaultValue(0);
+
+    $fields['password_enhancements_password_change_required'] = BaseFieldDefinition::create('boolean')
+      ->setLabel(t('Password change required'))
+      ->setDescription(t('Forces the user to change the password.<br><strong>Warning:</strong> This also causes to force logout the user.'))
+      ->setDefaultValue(0)
+      ->setDisplayOptions('form', [
+        'type' => 'boolean_checkbox',
+        'settings' => [
+          'display_label' => TRUE,
+        ],
+        'weight' => 99,
+      ]);
+  }
+
+  return $fields;
+}
+
+/**
+ * Implements hook_ENTITY_TYPE_presave().
+ */
+function password_enhancements_user_presave(EntityInterface $entity) {
+  // Update the last password changed date if the password was changed.
+  if ($entity->bundle() == 'user' && (empty($entity->original) || $entity->original->pass->value != $entity->pass->value)) {
+    $entity->password_enhancements_password_changed_date->setValue(Drupal::time()->getRequestTime());
+  }
+}
+
+/**
+ * Implements hook_user_login().
+ */
+function password_enhancements_user_login(UserInterface $user): void {
+  // Load password the password policy with the highest priority for the given
+  // user.
+  /** @var \Drupal\password_enhancements\Entity\Storage\PolicyEntityStorageInterface $password_policy_storage */
+  $password_policy_storage = Drupal::entityTypeManager()->getStorage('password_enhancements_policy');
+  $policy = $password_policy_storage->loadByRoleAndPriority($user->getRoles());
+
+  /** @var \Drupal\Core\Config\Config $config */
+  $config = Drupal::getContainer()->get('config.factory')->get('password_enhancements.settings');
+
+  // Redirect the user to the password change page if the password was expired
+  // or if it was required by the admin.
+  $expire_seconds = !empty($policy) ? $policy->getExpireDays() * 86400 : 0;
+  $request_time = Drupal::time()->getRequestTime();
```

Comment:
But why? The variables should be defined just before they will be used, otherwise you could end up with defining a lot of variables before anything would use it and the immediate statement would stop execution because of an error, or a condition is not met.

I would rather not define every variable on the top of the function it is not C either :)